### PR TITLE
Modified Activity History Stats

### PR DESCRIPTION
**Pull request by toyhammered**
1. Implement Recalculate for Anime/Manga
2. Implement Increment/Decrement

### CODE REVIEW

File: app/models/concerns/stat/activity_history.rb [hunk 1]
```diff
@@ -1,26 +1,95 @@
...
-
-      # clear stats_data
       self.stats_data = {}
-      self.stats_data = {
-        total: activity.count,
-        activity: activity # collection of events
-      }
+
+      date_trunc = "date_trunc('day', library_events.created_at)"
+      progress = "((changed_data#>>'{progress,1}')::integer - (changed_data#>>'{progress,0}')::integer)"
```
[the remark this answers is not in the document]
Comment (wopian) [reply app/models/concerns/stat/activity_history.rb]: 😱

**Comment by toyhammered** [comment]
ready for review @NuckChorris

**Review by wopian — CHANGES_REQUESTED** [review]

**Review by wopian — CHANGES_REQUESTED** [review]

File: app/models/concerns/stat/activity_history.rb [hunk 2]
```diff
@@ -2,59 +2,212 @@ class Stat < ApplicationRecord
...
+
+    # both dates will be overwritten, just needed the format
     DEFAULT_STATS = {
-      'total' => 0,
-      'activity' => []
+      'days' => {},
+      'first_event_date' => 5.years.ago,
+      'last_update_date' => 5.years.ago,
+      'week_high_score' => 0,
+      'total_progress' => 0
```
Comment (NuckChorris) [inline app/models/concerns/stat/activity_history.rb:22]: Is there a reason for these not to be nil

File: app/models/concerns/stat/activity_history.rb [hunk 3]
```diff
@@ -2,59 +2,212 @@ class Stat < ApplicationRecord
...
       self.stats_data = {}
-      self.stats_data = {
-        total: activity.count,
-        activity: activity # collection of events
-      }
+
+      date_trunc = "date_trunc('day', library_events.created_at)"
+      progress0 = "(changed_data#>>'{progress,0}')::integer"
+      progress1 = "(changed_data#>>'{progress,1}')::integer"
+      progress = "(#{progress1} - #{progress0})"
```
Comment (NuckChorris) [inline app/models/concerns/stat/activity_history.rb:31]: How about `progress_old`, `progress_new`, and `progress_diff`

File: app/models/concerns/stat/activity_history.rb [hunk 4]
```diff
@@ -2,59 +2,212 @@ class Stat < ApplicationRecord
...
-                     .eager_load(media_column)
-
-      # clear stats_data
       self.stats_data = {}
-      self.stats_data = {
-        total: activity.count,
-        activity: activity # collection of events
-      }
+
+      date_trunc = "date_trunc('day', library_events.created_at)"
```
Comment (NuckChorris) [inline app/models/concerns/stat/activity_history.rb:28]: How about `creation_date`

File: app/models/concerns/stat/activity_history.rb [hunk 5]
```diff
@@ -2,59 +2,212 @@ class Stat < ApplicationRecord
...
+      progress0 = "(changed_data#>>'{progress,0}')::integer"
+      progress1 = "(changed_data#>>'{progress,1}')::integer"
+      progress = "(#{progress1} - #{progress0})"
+
+      case media_column
+      when :anime then watch_time = "sum(#{progress} * anime.episode_length)"
+      when :manga then watch_time = 0
+      end
+
+      activity_dates = user.library_events.where("changed_data ? 'progress'")
```
Comment (NuckChorris) [inline app/models/concerns/stat/activity_history.rb:38]: You should add a comment explaining what this is building, what this beast is gonna return

File: app/models/concerns/stat/activity_history.rb [hunk 6]
```diff
@@ -2,59 +2,212 @@ class Stat < ApplicationRecord
...
+        stats_data['total_progress'] += amount
+      end
+      # Set the week_high_score from their first library_event date
+      stats_data['week_high_score'] = week_high_score(activity_dates.first[0])
 
       save!
     end
 
+    def generate_missing_days(start_date)
+      # not happy about having to use Time.now but I think I have to...
```
Comment (NuckChorris) [inline app/models/concerns/stat/activity_history.rb:67]: Does `Date.today` not work?

**Review by NuckChorris — COMMENTED** [review]
Frankly I'm finding this logic really hard to follow.  I know I was the one who suggested trying this structure originally, but would it be better to just do a flat hash of `'2017-08-20' => { progress: 6, time: 138 }`

I had hoped this structure would make logic easier but it seems like I was wrong on that

**Comment by NuckChorris** [comment]
(Integrated into #310)
